File: Support/lib/ML/DType.cpp
//===----------------------------------------------------------------------===//
// Copyright (c) 2026, Modular Inc. All rights reserved.
//
// Licensed under the Apache License v2.0 with LLVM Exceptions:
// https://llvm.org/LICENSE.txt
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//===----------------------------------------------------------------------===//

#include "Support/ML/DType.h"
#include "Support/Error.h"
#include "Support/ErrorOr.h"
#include "Support/LLVMForwardDecls.h"
#include "Support/LogicalResult.h"
#include "llvm/ADT/APFloat.h"
#include "llvm/ADT/StringExtras.h"
#include "llvm/Support/MathExtras.h"
#include "llvm/Support/raw_ostream.h"
#include <cassert>
#include <climits>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <string>
#include <sys/types.h>
#include <utility>
using namespace M;

/// Returns true if the type is a float8 type.
bool DType::isFloat8() const { return isFloat() && getWidthInBits() == 8; }

/// Return the width of this element in bits.  This returns -1 for unknown
/// width values.
ssize_t DType::getWidthInBits() const {
  // Handle complex separately from per-element types below.  We know that
  // complex element types are always at least a byte in size.
  if (isComplex()) {
    ssize_t strippedWidth = stripComplex().getWidthInBits();
    if (strippedWidth == -1)
      return -1;
    return strippedWidth * 2;
  }

  // This switch handles special cases inline, or determines the logarithmic
  // size of each element and breaks for the overflow check.
  switch (getValue()) {
  default:
    if (isInt())
      return getIntegerWidthInBits();
    if (auto *semantics = getFloatSemantics())
      return APFloat::getSizeInBits(*semantics);
    return -1;

    // Handle other types.
  case DType::kBool:
    return 8;
  }
}

/// Return the in-memory size for an array of the specified type with the
/// specified number of elements, or -1 for non-numeric types or too large
/// values.  This supports densely packed sub-byte types like i1, i2, i4.
ssize_t DType::getSizeInBytes(size_t numElements) const {
  auto sizeOr = getSizeInBytesChecked(numElements);
  if (failed(sizeOr))
    return -1;
  return *sizeOr;
}

/// Return the in-memory size for an array of the specified type with the
/// specified number of elements, or Error for non-numeric types or too large
/// values.  This supports densely packed sub-byte types like i1, i2, i4.
FailureOr<size_t> DType::getSizeInBytesChecked(size_t numElements) const {
  // Handle complex separately from per-element types below.
  if (isComplex()) {
    if (numElements > std::numeric_limits<ssize_t>::max() >> 1)
      return failure();
    return stripComplex().getSizeInBytesChecked(numElements * 2);
  }

  // This switch handles special cases inline, or determines the logarithmic
  // size of each element and breaks for the overflow check.
  size_t widthShift;
  if (isInt()) {
    // For integers, we just return the bit-width turned into bytes.  We treat
    // i1/i2/i4 types as being a single byte.
    widthShift = getIntegerWidthInLogBits();
  } else if (isFloat()) {
    ssize_t bitCount = getWidthInBits();
    if (!llvm::isPowerOf2_32(bitCount)) {
      if (numElements > std::numeric_limits<size_t>::max() / bitCount)
        return failure();
      return llvm::divideCeil(numElements * bitCount, 8);
    }
    widthShift = llvm::Log2_32(bitCount);
  } else if (isBool()) {
    widthShift = 3; // kBool is stored in a single byte each.
  } else {
    // Unhandled dtype
    return failure();
  }

  // i1,i2,i4,fp4 values are packed densely in memory.
  // We're going to do a truncating division (with a shift right) by the
  // element size, so make sure we round up to the next byte.
  if (widthShift <= 3) {
    if (numElements > std::numeric_limits<size_t>::max() >> widthShift)
      return failure();
    return llvm::divideCeil(numElements << widthShift, 8);
  }

  auto byteShift = widthShift - 3;
  // Otherwise we're growing. Convert to byte shift amount.
  if (numElements > std::numeric_limits<ssize_t>::max() >> byteShift)
    return failure();
  return numElements << byteShift;
}

/// Return a complex type if it is valid, otherwise fail.
FailureOr<DType> DType::getComplexChecked(DType eltType) {
  if (eltType.getWidthInBits() < 8 || eltType.isComplex())
    return failure();
  return getComplex(eltType);
}

/// This turns the printed form of a dtype back into a DType or
/// returns None if it is an unrecognized name.
FailureOr<DType> DType::getFromString(StringRef str) {
  if (str.empty())
    return failure();
  switch (str[0]) {
  case 'f':
#define DECLARE_FLOAT(SHORT_NAME, ...)                                         \
  if (str == #SHORT_NAME)                                                      \
    return DType(DType::SHORT_NAME);
#include "Support/ML/FloatTypes.def"
#undef DECLARE_FLOAT
    return failure();
  case 'u':
  case 's':
    if (str.size() >= 3 && str[1] == 'i') {
      unsigned width = 0;
      if (str.drop_front(2).getAsInteger(10, width))
        return failure();
      return getInt(width, /*isSigned=*/str[0] == 's');
    }
    return failure();

  case 'b':
    if (str == "bool")
      return DType(kBool);
    // Handle the bf16 special case, since it's a floating point type which
    // does not start with the letter 'f'.
    if (str == "bf16")
      return DType(bf16);
    return failure();
  case 'c':
    if (str.starts_with("complex<") && str.back() == '>') {
      auto elt = getFromString(str.drop_front(8).drop_back());
      if (failed(elt))
        return failure();
      return getComplexChecked(*elt);
    }
    return failure();
  case 'i':
    if (str == "invalid")
      return DType(invalid);
    return failure();
  default:
    // TODO: Could handle the eltType<unknown42> syntax if we wanted to.
    return failure();
  }
}

/// Return a string form of this eltType suitable for printing and error
/// messages.
std::string DType::getAsString() const {
  if (isComplex())
    return "complex<" + stripComplex().getAsString() + ">";
  if (isUInt())
    return "ui" + llvm::utostr(getIntegerWidthInBits());
  if (isSInt())
    return "si" + llvm::utostr(getIntegerWidthInBits());

  switch (getValue()) {
#define DECLARE_FLOAT(SHORT_NAME, ...)                                         \
  case DType::SHORT_NAME:                                                      \
    return #SHORT_NAME;
#include "Support/ML/FloatTypes.def"
#undef DECLARE_FLOAT
  case kBool:
    return "bool";
  case invalid:
    return "invalid";
  default:
    return "eltType<unknown" + llvm::utostr(getValue()) + ">";
  }
}

void DType::print(raw_ostream &os) const { os << getAsString(); }
void DType::dump() const { print(llvm::errs()); }

ErrorOr<std::pair<int32_t, int32_t>> DType::getMaxAndMinValue() const {
  return dispatch<ErrorOr<std::pair<int32_t, int32_t>>>()
      .when<DType::si32>([&]() {
        return std::pair(std::numeric_limits<int32_t>::max(),
                         std::numeric_limits<int32_t>::min());
      })
      .when<DType::si16>([&]() {
        return std::pair(std::numeric_limits<int16_t>::max(),
                         std::numeric_limits<int16_t>::min());
      })
      .when<DType::ui16>([&]() {
        return std::pair(std::numeric_limits<uint16_t>::max(),
                         std::numeric_limits<uint16_t>::min());
      })
      .when<DType::si8>([&]() {
        return std::pair(std::numeric_limits<int8_t>::max(),
                         std::numeric_limits<int8_t>::min());
      })
      .when<DType::ui8>([&]() {
        return std::pair(std::numeric_limits<uint8_t>::max(),
                         std::numeric_limits<uint8_t>::min());
      })
      .otherwise([&]() {
        return Error("Unsupported quantization dtype " + getAsString());
      });
}

/// This method returns the LLVM floating point semantics for the given DType,
/// or nullptr if the DType is not a floating point type LLVM knows about
/// (e.g. TF32).
const llvm::fltSemantics *DType::getFloatSemantics() const {
  switch (getValue()) {
  default:
    return nullptr;

#define DECLARE_FLOAT(SHORT_NAME, LONG_NAME, M_TYPE, MLIR_TYPE, CXX_TYPE,      \
                      APFLOAT_TYPE, ...)                                       \
  case DType::SHORT_NAME:                                                      \
    return &APFLOAT_TYPE();
#include "Support/ML/FloatTypes.def"
#undef DECLARE_FLOAT
  }
}

size_t DType::getAlignment() const {
  constexpr auto getAlign = [](const auto *ptr) {
    return alignof(decltype(*ptr));
  };
  return dispatch<size_t>((void *)nullptr)
      .when<DType::f16>(getAlign)
      .when<DType::bf16>(getAlign)
      .when<DType::bf16>(getAlign)
      .when<DType::f4e2m1fn>(getAlign)
      .when<DType::f6e2m3fn>(getAlign)
      .when<DType::f6e3m2fn>(getAlign)
      .when<DType::f8e8m0fnu>(getAlign)
      .when<DType::f8e4m3fn>(getAlign)
      .when<DType::f8e4m3fnuz>(getAlign)
      .when<DType::f8e5m2>(getAlign)
      .when<DType::f8e5m2fnuz>(getAlign)
      .whenCXXType(getAlign);
}

File: max/python/max/dtype/dtype.py
# ===----------------------------------------------------------------------=== #
# Copyright (c) 2026, Modular Inc. All rights reserved.
#
# Licensed under the Apache License v2.0 with LLVM Exceptions:
# https://llvm.org/LICENSE.txt
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ===----------------------------------------------------------------------=== #
"""Data types for tensors in MAX Engine."""

from __future__ import annotations

from typing import Any

import numpy as np
from max._core.dtype import DType as DType


def _missing(value: Any) -> DType | None:
    if isinstance(value, str):
        return _MLIR_TO_DTYPE[value]
    return None


def _repr(self: DType) -> str:
    return self.name


def _mlir(self: DType) -> str:
    return _DTYPE_TO_MLIR[self]


_DTYPE_TO_NUMPY = {
    DType.bool: np.bool_,
    DType.int8: np.int8,
    DType.int16: np.int16,
    DType.int32: np.int32,
    DType.int64: np.int64,
    DType.uint8: np.uint8,
    DType.uint16: np.uint16,
    DType.uint32: np.uint32,
    DType.uint64: np.uint64,
    DType.float8_e8m0fnu: np.uint8,
    DType.float8_e4m3fn: np.uint8,
    DType.float8_e4m3fnuz: np.uint8,
    DType.float8_e5m2: np.uint8,
    DType.float8_e5m2fnuz: np.uint8,
    DType.float16: np.float16,
    DType.float32: np.float32,
    DType.float64: np.float64,
}

_NUMPY_TO_DTYPE = {
    np.bool_: DType.bool,
    np.int8: DType.int8,
    np.int16: DType.int16,
    np.int32: DType.int32,
    np.int64: DType.int64,
    np.uint8: DType.uint8,
    np.uint16: DType.uint16,
    np.uint32: DType.uint32,
    np.uint64: DType.uint64,
    np.float16: DType.float16,
    np.float32: DType.float32,
    np.float64: DType.float64,
}


def _to_numpy(self: DType) -> np.dtype[Any]:
    """Converts this ``DType`` to the corresponding NumPy dtype.

    Returns:
        DType: The corresponding NumPy dtype object.

    Raises:
        ValueError: If the dtype is not supported.
    """
    if numpy_dtype := _DTYPE_TO_NUMPY.get(self):
        return np.dtype(numpy_dtype)
    raise ValueError(f"unsupported DType to convert to NumPy: {self}")


def _from_numpy(dtype: np.dtype[Any]) -> DType:
    """Converts a NumPy dtype to the corresponding DType.

    Args:
        dtype: The NumPy dtype to convert.

    Returns:
        DType: The corresponding DType enum value.

    Raises:
        ValueError: If the input dtype is not supported.
    """
    # Handle both np.dtype objects and numpy type objects.
    np_type = dtype.type if isinstance(dtype, np.dtype) else dtype

    if max_dtype := _NUMPY_TO_DTYPE.get(np_type):
        return max_dtype
    raise ValueError(f"unsupported NumPy dtype: {dtype}")


_DTYPE_TO_MLIR = {
    DType.bool: "i1",
    DType.int8: "si8",
    DType.int16: "si16",
    DType.int32: "si32",
    DType.int64: "si64",
    DType.uint8: "ui8",
    DType.uint16: "ui16",
    DType.uint32: "ui32",
    DType.uint64: "ui64",
    DType.float4_e2m1fn: "f4e2m1fn",
    DType.float6_e2m3fn: "f6e2m3fn",
    DType.float6_e3m2fn: "f6e3m2fn",
    DType.float8_e8m0fnu: "f8e8m0fnu",
    DType.float8_e4m3fn: "f8e4m3fn",
    DType.float8_e4m3fnuz: "f8e4m3fnuz",
    DType.float8_e5m2: "f8e5m2",
    DType.float8_e5m2fnuz: "f8e5m2fnuz",
    DType.float16: "f16",
    DType.float32: "f32",
    DType.float64: "f64",
    DType.bfloat16: "bf16",
}

_MLIR_TO_DTYPE = {v: k for k, v in _DTYPE_TO_MLIR.items()}


DType._missing_ = _missing  # type: ignore[method-assign]
DType.__repr__ = _repr  # type: ignore[method-assign, assignment]
DType._mlir = property(_mlir)  # type: ignore[assignment]
DType.to_numpy = _to_numpy  # type: ignore[method-assign]
DType.from_numpy = _from_numpy  # type: ignore[method-assign]
